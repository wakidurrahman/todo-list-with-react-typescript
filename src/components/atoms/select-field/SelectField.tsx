/**
 * SelectField Component
 *
 * A reusable form select component that follows Bootstrap styling conventions
 * and implements accessibility best practices.
 *
 * Features:
 * - Single and multiple selection support
 * - Built-in validation states and feedback
 * - Help text and error message display
 * - Accessibility support with ARIA attributes
 * - Bootstrap form styling
 * - Customizable size and appearance
 * - Default option handling
 * - Disabled options support
 *
 * @example
 * <SelectField
 *   id="country"
 *   label="Country"
 *   options={[
 *     { value: 'us', label: 'United States' },
 *     { value: 'ca', label: 'Canada' }
 *   ]}
 *   value={selectedCountry}
 *   onChange={handleCountryChange}
 *   required
 * />
 */

import React from 'react';

type SelectSize = 'sm' | 'lg' | undefined;

type SelectOption = {
  /** Value of the option */
  value: string;
  /** Display label for the option */
  label: string;
  /** Whether the option is disabled */
  disabled?: boolean;
};

type SelectFieldProps = {
  /** Options for the select */
  options: SelectOption[];
  /** Selected value */
  value: string;
  /** Callback when selection changes */
  onChange: (e: React.ChangeEvent<HTMLSelectElement>) => void;
  /** Select label */
  label?: string;
  /** Select id - required for accessibility */
  id: string;
  /** Select size - sm or lg */
  size?: SelectSize;
  /** Whether the select is disabled */
  disabled?: boolean;
  /** Help text below the select */
  helpText?: string;
  /** Whether the select is required */
  required?: boolean;
  /** Select name attribute */
  name?: string;
  /** Error message */
  error?: string;
  /** Additional CSS classes */
  className?: string;
  /** Aria-label for accessibility */
  'aria-label'?: string;
  /** Whether multiple options can be selected */
  multiple?: boolean;
  /** Size attribute for showing multiple options */
  visibleOptions?: number;
  /** Default option text */
  defaultOption?: string;
};

/**
 * SelectField component that renders a select element with associated label,
 * help text, and validation feedback.
 */
const SelectField = ({
  options,
  value,
  onChange,
  label,
  id,
  size,
  disabled = false,
  helpText,
  required = false,
  name,
  error,
  className = '',
  'aria-label': ariaLabel,
  multiple = false,
  visibleOptions,
  defaultOption = 'Select an option',
}: SelectFieldProps) => {
  // Generate unique IDs for accessibility elements
  const helpTextId = `${id}-help-text`;
  const errorId = `${id}-error`;

  // Combine aria-describedby values for accessibility
  const ariaDescribedbyValues = [
    helpText ? helpTextId : '',
    error ? errorId : '',
  ]
    .filter(Boolean)
    .join(' ');

  // Build class names for form select with validation states
  const selectClasses = [
    'form-select',
    size && `form-select-${size}`,
    error && 'is-invalid',
    className,
  ]
    .filter(Boolean)
    .join(' ');

  return (
    <div className="mb-3">
      {label && (
        <label htmlFor={id} className="form-label">
          {label}
          {required && <span className="text-danger ms-1">*</span>}
        </label>
      )}
      <select
        className={selectClasses}
        id={id}
        name={name || id}
        value={value}
        onChange={onChange}
        disabled={disabled}
        required={required}
        aria-label={ariaLabel}
        aria-describedby={ariaDescribedbyValues || undefined}
        multiple={multiple}
        size={visibleOptions}
      >
        {!multiple && (
          <option value="" disabled>
            {defaultOption}
          </option>
        )}
        {options.map((option) => (
          <option
            key={option.value}
            value={option.value}
            disabled={option.disabled}
          >
            {option.label}
          </option>
        ))}
      </select>

      {/* Help text */}
      {helpText && (
        <div id={helpTextId} className="form-text">
          {helpText}
        </div>
      )}

      {/* Error message */}
      {error && (
        <div id={errorId} className="invalid-feedback">
          {error}
        </div>
      )}
    </div>
  );
};

export default SelectField;
