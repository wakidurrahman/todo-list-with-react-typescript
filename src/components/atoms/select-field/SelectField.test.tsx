import '@testing-library/jest-dom';
import { fireEvent, render, screen } from '@testing-library/react';
import renderer from 'react-test-renderer';

import SelectField from './SelectField';

/**
 * Test suite for the SelectField component
 *
 * Categories:
 * - Basic Rendering
 *   - Default select rendering
 *   - Options rendering
 *   - Label rendering
 *   - Default option
 *
 * - Select Behavior
 *   - Value selection
 *   - Multiple selection
 *   - Disabled state
 *   - Disabled options
 *
 * - Validation and Feedback
 *   - Error messages
 *   - Help text
 *   - Required field
 *
 * - Accessibility
 *   - ARIA attributes
 *   - Label associations
 *
 * - Styling
 *   - Size variants
 *   - Custom classes
 */
describe('SelectField Component', () => {
  const defaultOptions = [
    { value: 'option1', label: 'Option 1' },
    { value: 'option2', label: 'Option 2' },
    { value: 'option3', label: 'Option 3' },
  ];

  const defaultProps = {
    id: 'test-select',
    options: defaultOptions,
    value: '',
    onChange: jest.fn(),
  };

  afterEach(() => {
    jest.clearAllMocks();
  });

  // Basic rendering tests
  describe('Basic Rendering', () => {
    it('renders select field with default props', () => {
      render(<SelectField {...defaultProps} />);
      expect(screen.getByRole('combobox')).toBeInTheDocument();
    });

    it('renders all options correctly', () => {
      render(<SelectField {...defaultProps} />);
      defaultOptions.forEach((option) => {
        expect(screen.getByText(option.label)).toBeInTheDocument();
      });
    });

    it('renders label when provided', () => {
      const label = 'Test Label';
      render(<SelectField {...defaultProps} label={label} />);
      expect(screen.getByText(label)).toBeInTheDocument();
    });

    it('renders default option', () => {
      const defaultOption = 'Choose an option';
      render(<SelectField {...defaultProps} defaultOption={defaultOption} />);
      expect(screen.getByText(defaultOption)).toBeInTheDocument();
    });
  });

  // Select behavior tests
  describe('Select Behavior', () => {
    it('handles value changes', () => {
      const handleChange = jest.fn();
      render(<SelectField {...defaultProps} onChange={handleChange} />);

      fireEvent.change(screen.getByRole('combobox'), {
        target: { value: 'option1' },
      });

      expect(handleChange).toHaveBeenCalledTimes(1);
    });

    it('supports multiple selection', () => {
      render(<SelectField {...defaultProps} multiple />);
      const select = screen.getByRole('listbox');
      expect(select).toHaveAttribute('multiple');
    });

    it('respects disabled state', () => {
      render(<SelectField {...defaultProps} disabled />);
      expect(screen.getByRole('combobox')).toBeDisabled();
    });

    it('handles disabled options', () => {
      const optionsWithDisabled = [
        ...defaultOptions,
        { value: 'disabled', label: 'Disabled Option', disabled: true },
      ];

      render(<SelectField {...defaultProps} options={optionsWithDisabled} />);
      const disabledOption = screen.getByText('Disabled Option');
      expect(disabledOption.closest('option')).toHaveAttribute('disabled');
    });
  });

  // Validation and feedback tests
  describe('Validation and Feedback', () => {
    it('shows error message when provided', () => {
      const errorMessage = 'This field is required';
      render(<SelectField {...defaultProps} error={errorMessage} />);
      expect(screen.getByText(errorMessage)).toHaveClass('invalid-feedback');
    });

    it('shows help text when provided', () => {
      const helpText = 'Select your option';
      render(<SelectField {...defaultProps} helpText={helpText} />);
      expect(screen.getByText(helpText)).toHaveClass('form-text');
    });

    it('shows required indicator when required', () => {
      render(<SelectField {...defaultProps} label="Test Label" required />);
      expect(screen.getByText('*')).toHaveClass('text-danger');
    });

    it('applies validation classes correctly', () => {
      render(<SelectField {...defaultProps} error="Error" />);
      expect(screen.getByRole('combobox')).toHaveClass('is-invalid');
    });
  });

  // Accessibility tests
  describe('Accessibility', () => {
    it('associates help text with select via aria-describedby', () => {
      const helpText = 'Help text';
      render(<SelectField {...defaultProps} helpText={helpText} />);

      const select = screen.getByRole('combobox');
      const helpTextElement = screen.getByText(helpText);

      expect(select).toHaveAttribute('aria-describedby', helpTextElement.id);
    });

    it('applies custom aria-label', () => {
      const ariaLabel = 'Custom label';
      render(<SelectField {...defaultProps} aria-label={ariaLabel} />);
      expect(screen.getByRole('combobox')).toHaveAttribute(
        'aria-label',
        ariaLabel
      );
    });

    it('associates label with select using htmlFor', () => {
      render(<SelectField {...defaultProps} label="Test Label" />);
      const select = screen.getByRole('combobox');
      const label = screen.getByText('Test Label');
      expect(label).toHaveAttribute('for', defaultProps.id);
      expect(select).toHaveAttribute('id', defaultProps.id);
    });
  });

  // Style tests
  describe('Styling', () => {
    it('applies size classes correctly', () => {
      render(<SelectField {...defaultProps} size="lg" />);
      expect(screen.getByRole('combobox')).toHaveClass('form-select-lg');
    });

    it('applies custom className', () => {
      const customClass = 'custom-select';
      render(<SelectField {...defaultProps} className={customClass} />);
      expect(screen.getByRole('combobox')).toHaveClass(customClass);
    });
  });

  // Snapshot tests
  describe('Snapshots', () => {
    it('matches basic select snapshot', () => {
      const tree = renderer
        .create(<SelectField {...defaultProps} label="Basic Select" />)
        .toJSON();
      expect(tree).toMatchSnapshot();
    });

    it('matches multiple select snapshot', () => {
      const tree = renderer
        .create(
          <SelectField
            {...defaultProps}
            multiple
            visibleOptions={3}
            label="Multiple Select"
          />
        )
        .toJSON();
      expect(tree).toMatchSnapshot();
    });

    it('matches select with validation snapshot', () => {
      const tree = renderer
        .create(
          <SelectField
            {...defaultProps}
            label="Validated Select"
            error="Error message"
            helpText="Help text"
            required
          />
        )
        .toJSON();
      expect(tree).toMatchSnapshot();
    });

    it('matches select with disabled options snapshot', () => {
      const optionsWithDisabled = [
        ...defaultOptions,
        { value: 'disabled', label: 'Disabled Option', disabled: true },
      ];

      const tree = renderer
        .create(
          <SelectField
            {...defaultProps}
            options={optionsWithDisabled}
            label="Select with Disabled Option"
          />
        )
        .toJSON();
      expect(tree).toMatchSnapshot();
    });
  });
});
