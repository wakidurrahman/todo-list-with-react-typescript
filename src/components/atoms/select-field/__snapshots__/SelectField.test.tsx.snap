// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SelectField Component Snapshots matches basic select snapshot 1`] = `
<div
  className="mb-3"
>
  <label
    className="form-label"
    htmlFor="test-select"
  >
    Basic Select
  </label>
  <select
    className="form-select"
    disabled={false}
    id="test-select"
    multiple={false}
    name="test-select"
    onChange={[MockFunction]}
    required={false}
    value=""
  >
    <option
      disabled={true}
      value=""
    >
      Select an option
    </option>
    <option
      value="option1"
    >
      Option 1
    </option>
    <option
      value="option2"
    >
      Option 2
    </option>
    <option
      value="option3"
    >
      Option 3
    </option>
  </select>
</div>
`;

exports[`SelectField Component Snapshots matches multiple select snapshot 1`] = `
<div
  className="mb-3"
>
  <label
    className="form-label"
    htmlFor="test-select"
  >
    Multiple Select
  </label>
  <select
    className="form-select"
    disabled={false}
    id="test-select"
    multiple={true}
    name="test-select"
    onChange={[MockFunction]}
    required={false}
    size={3}
    value=""
  >
    <option
      value="option1"
    >
      Option 1
    </option>
    <option
      value="option2"
    >
      Option 2
    </option>
    <option
      value="option3"
    >
      Option 3
    </option>
  </select>
</div>
`;

exports[`SelectField Component Snapshots matches select with disabled options snapshot 1`] = `
<div
  className="mb-3"
>
  <label
    className="form-label"
    htmlFor="test-select"
  >
    Select with Disabled Option
  </label>
  <select
    className="form-select"
    disabled={false}
    id="test-select"
    multiple={false}
    name="test-select"
    onChange={[MockFunction]}
    required={false}
    value=""
  >
    <option
      disabled={true}
      value=""
    >
      Select an option
    </option>
    <option
      value="option1"
    >
      Option 1
    </option>
    <option
      value="option2"
    >
      Option 2
    </option>
    <option
      value="option3"
    >
      Option 3
    </option>
    <option
      disabled={true}
      value="disabled"
    >
      Disabled Option
    </option>
  </select>
</div>
`;

exports[`SelectField Component Snapshots matches select with validation snapshot 1`] = `
<div
  className="mb-3"
>
  <label
    className="form-label"
    htmlFor="test-select"
  >
    Validated Select
    <span
      className="text-danger ms-1"
    >
      *
    </span>
  </label>
  <select
    aria-describedby="test-select-help-text test-select-error"
    className="form-select is-invalid"
    disabled={false}
    id="test-select"
    multiple={false}
    name="test-select"
    onChange={[MockFunction]}
    required={true}
    value=""
  >
    <option
      disabled={true}
      value=""
    >
      Select an option
    </option>
    <option
      value="option1"
    >
      Option 1
    </option>
    <option
      value="option2"
    >
      Option 2
    </option>
    <option
      value="option3"
    >
      Option 3
    </option>
  </select>
  <div
    className="form-text"
    id="test-select-help-text"
  >
    Help text
  </div>
  <div
    className="invalid-feedback"
    id="test-select-error"
  >
    Error message
  </div>
</div>
`;
